File: ResQNet_Project/resqnet-backend/src/main/java/com/pu_deltaforce/resqnet_backend/model/SafeZone.java
package com.pu_deltaforce.resqnet_backend.model;

import jakarta.persistence.*;
import lombok.Data;

@Entity
@Table(name = "safe_zones")
@Data
public class SafeZone {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String address;

    @Column(nullable = false)
    private Double latitude;

    @Column(nullable = false)
    private Double longitude;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ZoneType zoneType;

    private Integer capacity;
    private Integer currentOccupancy;
    private boolean available;

    public enum ZoneType {
        SHELTER, HOSPITAL, SCHOOL, MOSQUE, OTHER
    }

    @PrePersist
    public void prePersist() {
        this.available = true;
        if (this.currentOccupancy == null) this.currentOccupancy = 0;
    }
}

File: ResQNet_Project/resqnet-frontend/src/app/core/services/safezone.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../environments/environment';

export interface SafeZone {
  id?: number;
  name: string;
  address: string;
  zoneType: 'SHELTER' | 'HOSPITAL' | 'SCHOOL' | 'MOSQUE' | 'OTHER';
  capacity: number;
  currentOccupancy: number;
  available?: boolean;
  latitude: number;
  longitude: number;
  contactNumber: string;
}

@Injectable({ providedIn: 'root' })
export class SafezoneService {
  private apiUrl = `${environment.apiUrl}/safezones`;

  constructor(private http: HttpClient) {}

  getAll() {
    return this.http.get<SafeZone[]>(this.apiUrl);
  }

  create(data: SafeZone) {
    return this.http.post<SafeZone>(this.apiUrl, data);
  }

  updateOccupancy(id: number, occupancy: number) {
    return this.http.put(`${this.apiUrl}/${id}/occupancy?occupancy=${occupancy}`, {});
  }

  toggleAvailability(id: number) {
    return this.http.put(`${this.apiUrl}/${id}/toggle-availability`, {});
  }
}